sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -193,16 +193,43 @@ def _check_categories(self, X):
         if categorical_features.size == 0:
             return None, None
 
-        if categorical_features.dtype.kind not in ("i", "b"):
+        if categorical_features.dtype.kind not in ("i", "b", "U", "O"):
             raise ValueError(
-                "categorical_features must be an array-like of "
-                "bools or array-like of ints."
+                "categorical_features must be an array-like of bool, int or "
+                f"str, got: {categorical_features.dtype.name}."
             )
 
+        if categorical_features.dtype.kind == "O":
+            types = set(type(f) for f in categorical_features)
+            if types != {str}:
+                raise ValueError(
+                    "categorical_features must be an array-like of bool, int or "
+                    f"str, got: {', '.join(t.__name__ for t in types)}."
+                )
+
         n_features = X.shape[1]
 
-        # check for categorical features as indices
-        if categorical_features.dtype.kind == "i":
+        if categorical_features.dtype.kind in ("U", "O"):
+            # check for feature names
+            if not hasattr(self, "feature_names_in_"):
+                raise ValueError(
+                    "categorical_features should be passed as an array of "
+                    "integers or as a boolean mask when the model is fitted "
+                    "on data without feature names."
+                )
+            is_categorical = np.zeros(n_features, dtype=bool)
+            feature_names = self.feature_names_in_.tolist()
+            for feature_name in categorical_features:
+                try:
+                    is_categorical[feature_names.index(feature_name)] = True
```

Comment:
I am wondering if we should put in common this part with something that I implemented in the PDP part: https://github.com/scikit-learn/scikit-learn/pull/18298/files#diff-0239129ee36e71eb37218dd17745c41fba4b899793554791675d8a4bdaf81485R22-R48

Reply:
I mean that we could maybe anticipate having shared utils for categorical and feature names?

Reply:
I have this PR open in a tab for review, I swear :)

Reply:
Fine with me